File: leetsync/3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.go
func countSubarrays(nums []int, k int) int64 {
	n := len(nums)
	left, right, max, maxFreq := 0, 0, 0, 0
	result := 0
	for _, num := range nums {
		if max < num {
			max = num
		}
	}
	for right < n {
		if nums[right] == max {
			maxFreq++
		}
		for maxFreq >= k {
			result += n - right
			if nums[left] == max {
				maxFreq--
			}
			left++
		}
		right++
	}
	return int64(result)
}

File: leetsync/3213-count-subarrays-where-max-element-appears-at-least-k-times/count-subarrays-where-max-element-appears-at-least-k-times.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int n = nums.length;
        long result = 0;
        int max = Integer.MIN_VALUE;
        for (int num : nums) {
            if (max < num) {
                max = num;
            }
        }
        int left = 0, right = 0, maxFreq = 0;
        while (right < n) {
            if (nums[right] == max) {
                maxFreq++;
            }
            while (maxFreq >= k) {
                result += n - right;
                if (nums[left] == max) {
                    maxFreq--;
                }
                left++;
            }
            right++;
        }
        return result;
    }
}